libcudacxx/include/cuda/__span/shared_memory_span.h

```diff
@@ -0,0 +1,719 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDA___SPAN_SHARED_MEMORY_SPAN_H
+#define _CUDA___SPAN_SHARED_MEMORY_SPAN_H
+
+#include <cuda/std/detail/__config>
+
+#include <cstddef>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#if _CCCL_CUDA_COMPILATION()
+
+#  include <cuda/__memory/address_space.h>
+#  include <cuda/__ptx/instructions/get_sreg.h>
+#  include <cuda/std/__concepts/concept_macros.h>
+#  include <cuda/std/__exception/terminate.h>
+#  include <cuda/std/__iterator/reverse_iterator.h>
+#  include <cuda/std/__limits/numeric_limits.h>
+#  include <cuda/std/__memory/pointer_traits.h>
+#  include <cuda/std/__type_traits/is_const.h>
+#  include <cuda/std/__type_traits/is_constant_evaluated.h>
+#  include <cuda/std/__type_traits/remove_cv.h>
+#  include <cuda/std/array>
+#  include <cuda/std/cstddef>
+#  include <cuda/std/cstdint>
+#  include <cuda/std/initializer_list>
+#  include <cuda/std/span>
+
+#  include <cuda/std/__cccl/prologue.h>
+
+_CCCL_BEGIN_NAMESPACE_CUDA_DEVICE
+
+using __smem_size_t = ::cuda::std::uint32_t;
+
+template <typename _Tp>
+[[nodiscard]] _CCCL_DEVICE_API _Tp* __smem_to_pointer(__smem_size_t __raw_smem_ptr) noexcept
+{
+  auto __ptr = reinterpret_cast<_Tp*>(::__cvta_shared_to_generic(__raw_smem_ptr));
+  bool __p   = ::__isShared(__ptr);
+  _CCCL_ASSUME(__p);
+  return __ptr;
+}
+
+[[nodiscard]] _CCCL_DEVICE_API __smem_size_t __pointer_to_smem(const void* __smem_ptr) noexcept
+{
+  _CCCL_ASSERT(::cuda::device::is_address_from(__smem_ptr, ::cuda::device::address_space::shared),
+               "invalid smem pointer");
+  return static_cast<__smem_size_t>(::__cvta_generic_to_shared(__smem_ptr));
+}
+
+[[nodiscard]] _CCCL_DEVICE_API __smem_size_t __max_smem_allocation_bytes() noexcept
+{
+  const auto __total_smem_size   = ::cuda::ptx::get_sreg_total_smem_size();
+  const auto __dynamic_smem_size = ::cuda::ptx::get_sreg_dynamic_smem_size();
+  const auto __static_smem_size  = __total_smem_size - __dynamic_smem_size;
+  const auto __max_smem_size     = ::max(__static_smem_size, __dynamic_smem_size);
+  return __max_smem_size;
```

Comment:
Question: This makes me wonder how I can create a span over the SMEM of several CTAs in the same cluster. I assume this span's design does not support cluster/distributed SMEM (DSMEM)?

Reply:
ok, great!